cedar-drt/fuzz/fuzz_targets/roundtrip.rs

```diff
@@ -0,0 +1,201 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#![no_main]
+
+use cedar_drt::initialize_log;
+use cedar_drt_inner::fuzz_target;
+use cedar_policy_core::ast::{self, EntityType, ExprKind, Literal, StaticPolicy, Template};
+use cedar_policy_core::est;
+use cedar_policy_core::parser::{self, parse_policy};
+use cedar_policy_generators::{
+    abac::ABACPolicy, hierarchy::HierarchyGenerator, schema::Schema, settings::ABACSettings,
+};
+use libfuzzer_sys::arbitrary::{self, Arbitrary, Unstructured};
+use log::debug;
+use serde::Serialize;
+
+// A thin wrapper for policy
+#[derive(Debug, Clone, Serialize)]
+struct FuzzTargetInput {
+    // the generated policy
+    policy: ABACPolicy,
+}
+
+// settings for this fuzz target
+// copy-pasted from abac.rs
+const SETTINGS: ABACSettings = ABACSettings {
+    match_types: false,
+    enable_extensions: true,
+    max_depth: 7,
+    max_width: 7,
+    enable_additional_attributes: true,
+    enable_like: true,
+    enable_action_groups_and_attrs: true,
+    enable_arbitrary_func_call: false,
+    enable_unknowns: false,
+    enable_action_in_constraints: true,
+    enable_unspecified_apply_spec: true,
+};
+
+impl<'a> Arbitrary<'a> for FuzzTargetInput {
+    fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        let schema: Schema = Schema::arbitrary(SETTINGS.clone(), u)?;
+        let hierarchy = schema.arbitrary_hierarchy(u)?;
+        let policy = schema.arbitrary_policy(&hierarchy, u)?;
+        Ok(Self { policy })
+    }
+
+    fn size_hint(depth: usize) -> (usize, Option<usize>) {
+        arbitrary::size_hint::and_all(&[
+            Schema::arbitrary_size_hint(depth),
+            HierarchyGenerator::size_hint(depth),
+            Schema::arbitrary_policy_size_hint(&SETTINGS, depth),
+        ])
+    }
+}
+
+fn contains_unspecified_entities(p: &StaticPolicy) -> bool {
+    p.condition().subexpressions().any(|e| matches!(e.expr_kind(), ExprKind::Lit(Literal::EntityUID(euid)) if matches!(*euid.entity_type(), EntityType::Unspecified)))
+}
+
+// Print a policy to a string and parse it back using the standard AST parser.
+// Panics if parsing fails.
+fn round_trip_ast(p: &StaticPolicy) -> StaticPolicy {
+    parse_policy(None, &p.to_string()).unwrap_or_else(|err| {
+        panic!(
+            "Failed to round-trip AST: {:?}\nPretty printed form: {}\nParse error: {:?}\n",
+            p, p, err
+        )
+    })
+}
+
+// Print a policy to a string, parse it back using the EST parser, and convert to JSON.
+// Panics if any step fails. The resulting policy may be different from the input
+// (since the roundtrip is lossy), so this function only checks that the round-trip
+// succeeds without error.
```

Comment:
The lossy step is EST->AST or CST->AST, neither of which is included in this roundtrip I think?  (I believe this takes AST input and goes AST->text->EST->JSON?)  So I think we actually could check that the resulting policy matches the input

Reply:
EST->AST would be required here because `check_policy_equivalence` expects two ASTs.

But it would be possible to define this function for ESTs too. I'll update this PR later to include more variants, inspired by the helper functions in `est.rs`.

Reply:
Update: just looked at this again, and I think this step actually _should_ be lossless. I assumed the translation was supposed to be lossy since DRT failed when I compare the outputs. But looking at the failure again, it looks like it's [cedar#618](https://github.com/cedar-policy/cedar/issues/618), which should be resolved by [cedar#622](https://github.com/cedar-policy/cedar/pull/622). I'll add in the code to do the AST comparison, and test again once cedar#622 is finished.